src/plugins/es_ui_shared/public/request/use_request.ts

```diff
@@ -126,50 +94,44 @@ export const useRequest = <D = any, E = Error>(
     const response = await sendRequest<D, E>(httpClient, requestBody);
     const { data: serializedResponseData, error: responseError } = response;
 
-    // If an outdated request has resolved, DON'T update state, but DO allow the processData handler
-    // to execute side effects like update telemetry.
+    // If an outdated request has resolved, ignore its outdated data.
     if (isOutdatedRequest) {
       return { data: null, error: null };
     }
 
     setError(responseError);
-
+    // If there's an error, keep the data from the last request in case it's still useful to the user.
     if (!responseError) {
       const responseData = deserializer(serializedResponseData);
       setData(responseData);
     }
-
     setIsLoading(false);
     setIsInitialRequest(false);
 
     // If we're on an interval, we need to schedule the next request. This also allows us to reset
     // the interval if the user has manually requested the data, to avoid doubled-up requests.
-    scheduleRequest();
+    scheduleRequestRef.current!();
 
     return { data: serializedResponseData, error: responseError };
   };
 
   useEffect(() => {
-    sendRequestRef.current = _sendRequest;
-  }, [_sendRequest]);
-
-  const stringifiedQuery = useMemo(() => JSON.stringify(query), [query]);
+    sendRequestRef.current!();
 
-  useEffect(() => {
-    (sendRequestRef.current ?? _sendRequest)();
     // To be functionally correct we'd send a new request if the method, path, query or body changes.
     // But it doesn't seem likely that the method will change and body is likely to be a new
     // object even if its shape hasn't changed, so for now we're just watching the path and the query.
-  }, [path, stringifiedQuery]);
+  }, [path, query]);
```

Comment:
These are only dependencies that I think should have any effect on the hook behavior, at least at this point. If we were to change the refs above to `useCallback`, I'm concerned about introducing unwanted side effects. For example, converting `sendRequestRef` to `useCallback` means defining dependencies that duplicate those defined here (see below).

I'm open to exploring this in separate PRs, but I think for the context of this PR I'd like to avoid introducing any complexity that doesn't have a clear, deliberate purpose (e.g. fixing a bug or defining behavior more clearly). I think the current code makes it clear what will happen if a) dependencies change and b) `sendRequestRef` is called. If other people feel the same way then I think that's enough for me to consider this merge-worthy in its current form.

```ts
  const sendRequest = useCallback(async () => {
    // We don't clear error or data, so it's up to the consumer to decide whether to display the
    // "old" error/data or loading state when a new request is in-flight.
    setIsLoading(true);

    const requestBody = {
      path,
      method,
      query,
      body,
    };

    const response = await sendRequest<D, E>(httpClient, requestBody);
    const { data: serializedResponseData, error: responseError } = response;

    // If an outdated request has resolved, ignore its outdated data.
    if (isOutdatedRequest) {
      return { data: null, error: null };
    }

    setError(responseError);
    // If there's an error, keep the data from the last request in case it's still useful to the user.
    if (!responseError) {
      const responseData = deserializer(serializedResponseData);
      setData(responseData);
    }
    setIsLoading(false);
    setIsInitialRequest(false);

    // If we're on an interval, we need to schedule the next request. This also allows us to reset
    // the interval if the user has manually requested the data, to avoid doubled-up requests.
    scheduleRequestRef.current!();

    return { data: serializedResponseData, error: responseError };
  }, [body, deserializer, httpClient, isOutdatedRequest, method, path, query]);
```